Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h

```diff
@@ -2695,6 +2706,185 @@ auto AirIRGeneratorBase<Derived, ExpressionType>::addStructSet(ExpressionType st
+            emitLoadRTTFromObject(reference, tmpForRTT);
+            emitCheckOrBranchForCast(castKind, [&]() {
+                return makeBranchNotRTTKind(tmpForRTT, signature.expand().is<Wasm::ArrayType>() ? RTTKind::Array : RTTKind::Struct);
+            }, castFailure, falseBlock);
+        }
+
+        auto tmpForTargetRTT = self().gPtr();
+        append(Move, Arg::immPtr(m_info.rtts[heapType].get()), tmpForTargetRTT);
+        append(Branch32, Arg::relCond(MacroAssembler::Equal), tmpForRTT, tmpForTargetRTT);
+        if (castKind == CastKind::Cast)
+            m_currentBlock->setSuccessors(continuation, slowPath);
+        else
+            m_currentBlock->setSuccessors(trueBlock, slowPath);
+
+        m_currentBlock = slowPath;
+        auto tmpForCall = self().g32();
+        // FIXME: It may be worthwhile to JIT inline this in the future.
+        emitCCall(&operationWasmIsSubRTT, tmpForCall, tmpForRTT, tmpForTargetRTT);
+        emitCheckOrBranchForCast(castKind, [&]() {
+            return Inst(BranchTest32, nullptr, Arg::resCond(MacroAssembler::Zero), tmpForCall, tmpForCall);
+        }, castFailure, falseBlock);
+    }
+    }
+
+    append(Jump);
+    if (castKind == CastKind::Test) {
+        m_currentBlock->setSuccessors(trueBlock);
+        m_currentBlock = trueBlock;
+        append(Move, Arg::imm(1), result);
+        append(Jump);
+        m_currentBlock->setSuccessors(continuation);
+
+        m_currentBlock = falseBlock;
+        append(Move, Arg::imm(0), result);
+        append(Jump);
+    }
+
+    m_currentBlock->setSuccessors(continuation);
+    m_currentBlock = continuation;
+}
+
+template <typename Derived, typename ExpressionType>
+template <typename Branch, typename Generator>
+void AirIRGeneratorBase<Derived, ExpressionType>::emitCheckOrBranchForCast(CastKind kind, const Branch& makeBranch, const Generator& generator, BasicBlock* falseBlock)
+{
+    if (kind == CastKind::Cast)
+        emitCheck(makeBranch, generator);
+    else {
+        ASSERT(falseBlock);
+        BasicBlock* success = m_code.addBlock();
+        m_currentBlock->appendInst(makeBranch());
+        m_currentBlock->setSuccessors(falseBlock, success);
+        m_currentBlock = success;
+    }
+}
+
+template <typename Derived, typename ExpressionType>
+void AirIRGeneratorBase<Derived, ExpressionType>::emitLoadRTTFromFuncref(ExpressionType funcref, ExpressionType& result)
+{
+    auto* patch = addPatchpoint(B3::Int64);
+    patch->setGenerator([](CCallHelpers& jit, const B3::StackmapGenerationParams& params) {
+        jit.loadCompactPtr(CCallHelpers::Address(params[1].gpr(), WebAssemblyFunctionBase::offsetOfRTT()), params[0].gpr());
+    });
+    self().emitPatchpoint(patch, result, funcref);
+}
+
+template <typename Derived, typename ExpressionType>
+void AirIRGeneratorBase<Derived, ExpressionType>::emitLoadRTTFromObject(ExpressionType cell, ExpressionType& result)
+{
+    self().emitLoad(cell, WebAssemblyGCObjectBase::offsetOfRTT(), result);
```

Comment:
Both Air and B3 implementations for this got simpler, as it's now just a plain load instead of a patchpoint.